python/ray/experimental/channel/torch_tensor_nccl_channel.py

```diff
@@ -0,0 +1,222 @@
+            )
+        if value.device != self._device:
+            raise ValueError(
+                f"torch.Tensor must be on the default device: {self._device}"
+            )
+
+        if readers is not None:
+            raise NotImplementedError(
+                "TorchTensorNcclChannel.write() not supported for dynamically "
+                "passed readers."
+            )
+
+        # TODO: If there are multiple readers, can replace with a broadcast.
+        for rank in self._reader_ranks:
+            self._nccl_group.send(value, rank)
+
+    def begin_read(self) -> torch.Tensor:
+        buf = torch.zeros(self._typ.shape, dtype=self._typ.dtype, device=self._device)
+        self._nccl_group.recv(buf, self._writer_rank)
+        return buf
+
+    def end_read(self) -> None:
+        return
+
+    def close(self) -> None:
+        self._nccl_group.destroy()
+
+
+def _do_init_nccl_group(self, world_size, comm_id, rank, actor_ids_to_ranks):
+    # assert ray.get_gpu_ids()
+
+    ctx = ChannelContext.get_current()
+    ctx.nccl_group = _NcclGroup(
+        world_size,
+        comm_id,
+        rank,
+        actor_ids_to_ranks,
+        torch.cuda.current_stream().cuda_stream,
+    )
+
+
+@DeveloperAPI
+def _do_check_has_gpu(self) -> bool:
+    return bool(ray.get_gpu_ids())
+
+
+def _init_nccl_group(
+    actors: List[ray.actor.ActorHandle],
+):
+    ctx = ChannelContext.get_current()
+    if ctx.nccl_group is not None:
+        return
+
+    has_gpus = ray.get(
+        [actor.__ray_call__.remote(_do_check_has_gpu) for actor in actors]
+    )
+    for has_gpu, actor in zip(has_gpus, actors):
+        if not has_gpu:
+            raise ValueError(
+                f"Actor {actor} returns a tensor with type hint "
+                'TorchTensor(transport="nccl") but actor does not have a '
+                "GPU assigned by Ray."
+            )
+
+    actor_handles_to_ranks = {actor: rank for rank, actor in enumerate(actors)}
+    actor_ids_to_ranks = {
+        actor._ray_actor_id: rank for actor, rank in actor_handles_to_ranks.items()
+    }
+    assert len(actor_ids_to_ranks) == len(
+        actor_handles_to_ranks
+    ), "actors must be unique"
+
+    from cupy.cuda import nccl
+
+    comm_id = nccl.get_unique_id()
+
+    logger.info(f"Creating NCCL group on actors->ranks: {actor_handles_to_ranks}")
+
+    world_size = len(actors)
+    init_tasks = [
+        actor.__ray_call__.remote(
+            _do_init_nccl_group,
+            world_size,
+            comm_id,
+            rank,
+            actor_ids_to_ranks,
+        )
+        for actor, rank in actor_handles_to_ranks.items()
+    ]
+    done = False
+    try:
+        ray.get(init_tasks, timeout=30)
+        done = True
+    except ray.exceptions.GetTimeoutError:
+        logger.warning(
+            "NCCL group creation not done after 30s. NCCL group creation may be hung."
+        )
+    if not done:
+        ray.get(init_tasks)
```

Comment:
This code seems redundant? Maybe we can just re-raise exception within except ray.exceptions.GetTimeoutError:?

Reply:
What do you mean exactly? The second ray.get blocks until the tasks complete or an actual error occurs, if the first one timed out.